### PR TITLE
Adding LanguagePrimitives.TryCompare to provide faster comparisons

### PR DESCRIPTION
**Pull request by powercode**
## PR Summary

This PR adds the  `TryCompare` equivalent of `Compare`, just as we have a `TryConvert` alternative to `Convert`.
This enables faster comparison in the cases where the convertions throws exception.

As an example:

```powershell
 1..10000 | foreach{$_; [timespan]::FromSeconds($_)}  | sort
```
takes `3 800 ms` before and `280 ms` after.

```powershell
$b = 1..1000000 | %{100; [timespan]::FromSeconds(100)}  | group
```
takes `20 s` before and `15 s` after.

It is implemented equivalently to `LanguagePrimitives.Compare`, but returning false where `Compare` would have thrown an exception.

This is adding a new public API to `LanguagePrimitives`, so it will probably need a committee approval.

The new APIs are:

```CSharp
class LanguagePrimitives {
  public static bool TryCompare(object first, object second, out int result);
  public static bool TryCompare(object first, object second, bool ignoreCase, out int result);
  /// <summary>
  /// Tries to compare first and second, converting second to the type of the first, if necessary.
  /// If a conversion is needed but fails, false is return.
  /// </summary>
  /// <param name="first">First comparison value.</param>
  /// <param name="second">Second comparison value.</param>
  /// <param name="ignoreCase">Used if both values are strings.</param>
  /// <param name="formatProvider">Used in type conversions and if both values are strings.</param>
  /// <param name="result">Less than zero if first is smaller than second, more than  zero if it is greater or zero if they are the same.</param>
  /// <returns>True if the comparison was successful, false otherwise.</returns>
  /// <exception cref="ArgumentException">The parameter <paramref name="formatProvider"/> is not a <see cref="CultureInfo"/>.</exception>
  public static bool TryCompare(
      object first, 
      object second, 
      bool ignoreCase, 
      IFormatProvider formatProvider,
      out int result);
}
```

The new APIs are used by `ObjectCommandPropertyValue` (used by sort-object and group-object) and `FormatGroupManager`.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

**Comment by powercode** [comment]
The remaining CodeFactor issues are intentionally left that way to match the style of similar code in the same file.

**Comment by iSazonov** [comment]
> This PR adds the Compare equivalent of TryConvert.

Seem typo in the PR description.

File: test/powershell/engine/Api/LanguagePrimitive.Tests.ps1 [hunk 1]
```diff
@@ -46,4 +46,60 @@ Describe "Language Primitive Tests" -Tags "CI" {
         [System.Management.Automation.LanguagePrimitives]::GetEnumerable($dt) | ForEach-Object { $count++ }
         $count | Should -Be 2
     }
+
+    It "TryCompare should succeed on int and string" {
+        $result = $null
+        [System.Management.Automation.LanguagePrimitives]::TryCompare(1, "1", [ref] $result) | Should -BeTrue
+        $result | Should -BeExactly 0
```
Comment (iSazonov) [inline test/powershell/engine/Api/LanguagePrimitive.Tests.ps1]: We use `-BeExactly` for string only. Here we expect an integer and should use `-Be`.
Please review using `-BeExactly` below too.

File: src/System.Management.Automation/engine/LanguagePrimitives.cs [hunk 2]
```diff
@@ -854,6 +854,146 @@ public static int Compare(object first, object second, bool ignoreCase, IFormatP
...
+        /// <exception cref="ArgumentException">The parameter <paramref name="formatProvider"/> is not a <see cref="CultureInfo"/>.</exception>
+        public static bool TryCompare(object first, object second, bool ignoreCase, IFormatProvider formatProvider, out int result)
+        {
+            result = 0;
+            if (formatProvider == null)
+            {
+                formatProvider = CultureInfo.InvariantCulture;
+            }
+
+            if (!(formatProvider is CultureInfo culture))
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/LanguagePrimitives.cs]: Why we use `IFormatProvider` parameter type if we want get CultureInfo?

File: src/System.Management.Automation/engine/LanguagePrimitives.cs [hunk 3]
```diff
@@ -854,6 +854,146 @@ public static int Compare(object first, object second, bool ignoreCase, IFormatP
...
+            {
+                throw PSTraceSource.NewArgumentException("formatProvider");
+            }
+
+            first = PSObject.Base(first);
+            second = PSObject.Base(second);
+
+            int CompareNullToNumeric(object value, bool numberIsRightHandSide)
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/LanguagePrimitives.cs]: I think we should use "dry" principle and use the function everywhere in the file.

File: src/System.Management.Automation/engine/LanguagePrimitives.cs [hunk 4]
```diff
@@ -854,6 +854,146 @@ public static int Compare(object first, object second, bool ignoreCase, IFormatP
...
+                // for everything else it's less than zero...
+                result = CompareNullToNumeric(first, false);
+                return true;
+            }
+
+            if (first is string firstString)
+            {
+                if (!(second is string secondString))
+                {
+                    if (!TryConvertTo(second, culture, out secondString))
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/LanguagePrimitives.cs]: We could replace the two if-s with one.

**Review by iSazonov — COMMENTED** [review]
Please add some samples that we really get a performance win.

New public API must be approved by @powershell-committee.

**Comment by SteveL-MSFT** [comment]
@PowerShell/powershell-committee reviewed this and is fine with the api additions.  Thanks.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ObjectCommandComparer.cs [hunk 5]
```diff
@@ -190,35 +189,30 @@ public int Compare(object first, object second)
...
-            }
-            catch (InvalidCastException)
-            {
-            }
-            catch (ArgumentException)
+            if (LanguagePrimitives.TryCompare(first, second, !_caseSensitive, _cultureInfo, out int result))
             {
-                // Note that this will occur if the objects do not support
-                // IComparable.  We fall back to comparing as strings.
+                return result * (_ascendingOrder ? 1 : -1);
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/ObjectCommandComparer.cs]: Do we get some win if we exclude multiplication?

File: src/System.Management.Automation/engine/LanguagePrimitives.cs [hunk 6]
```diff
@@ -854,6 +847,129 @@ public static int Compare(object first, object second, bool ignoreCase, IFormatP
...
+            {
+                throw PSTraceSource.NewArgumentException("formatProvider");
+            }
+
+            first = PSObject.Base(first);
+            second = PSObject.Base(second);
+
+
+
+            if (first == null && second == null)
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/LanguagePrimitives.cs]: Please remove extra empty lines above.

**Review by iSazonov — APPROVED** [review]
LGTM with 3 minor comments.

Thanks for the great performance improvements.

**Review by daxian-dbw — APPROVED** [review]
LGTM

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ObjectCommandComparer.cs (hunk 5 above)
Comment (powercode) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/ObjectCommandComparer.cs]: I'd leave that to the compiler and the JIT. They usually do a good job on things like this.

File: src/System.Management.Automation/engine/LanguagePrimitives.cs [hunk 7]
```diff
@@ -854,6 +847,127 @@ public static int Compare(object first, object second, bool ignoreCase, IFormatP
...
+                // for everything else it's less than zero...
+                result = CompareObjectToNull(first, false);
+                return true;
+            }
+
+            if (first is string firstString)
+            {
+                if (!(second is string secondString))
+                {
+                    if (!TryConvertTo(second, culture, out secondString))
```
[the remark this answers is not in the document]
Comment (powercode) [reply src/System.Management.Automation/engine/LanguagePrimitives.cs]: @PaulHigin I recently changed TryConvertTo to check the result of FigureConvertion and exit early, without invoking ConversionData.Invoke() if the ConvertionRank is None.

So, yes, I have avoided the exception in many but not all cases. Interesting take to put it on ConversionData. That is arguably where it belongs.

**Pull request merged**